### PR TITLE
moved __call__ arguments to new dispatch method

### PR DESCRIPTION
**Pull request by ordirules**
response to #2366 
Allow the lookup of the instance of `normalize_token` depending on class. Similar to `dispatch` from `functools.singledispatch`

### CODE REVIEW

File: dask/utils.py [hunk 1]
```diff
@@ -399,14 +402,21 @@ def __call__(self, arg):
             pass
         else:
             register()
-            return self(arg)  # recurse
+            return self  # recurse
```
Comment (jcrist) [inline dask/utils.py]: Should be `self.dispatch(typ)`

File: dask/tests/test_utils.py [hunk 2]
```diff
@@ -49,16 +49,17 @@ def test_dispatch():
     foo.register(float, lambda a: a - 1)
     foo.register(tuple, lambda a: tuple(foo(i) for i in a))
     foo.register(object, lambda a: a)
+    foo_int = foo.dispatch(int)
 
     class Bar(object):
         pass
     b = Bar()
     assert foo(1) == 2
+    assert foo_int(1) == 2
```
Comment (jcrist) [inline dask/tests/test_utils.py]: Preference to do move the dispatch call to where it's used in the test. `assert foo.dispatch(int)(1) == 2`

File: dask/utils.py [hunk 3]
```diff
@@ -381,16 +381,19 @@ def wrapper(func):
 
         return wrapper(func) if func is not None else wrapper
 
-    def __call__(self, arg):
+    def dispatch(self, typ):
+        """
+        Return the correct version of the dispatch function.
+        typ : the class type
```
Comment (jcrist) [inline dask/utils.py]: Request to rename `typ` to `cls`, and have the docstring be:

```
"""Return the function implementation for the given ``cls``"""
```

File: dask/utils.py (hunk 1 above)
Comment (ordirules) [inline dask/utils.py]: I just want to confirm, either case works correct? (logically to me looks fine) I even tested with `register_lazy` and I don't see an issue.
Basically, I ask this because if there is a logic difference, then maybe it's a good opportunity to add a test, thanks!

**Comment by ordirules** [comment]
thanks @jcrist made changes. I added a test for `register_lazy` as well since I didn't see one, while I was on that code. just a suggestion, I can remove it
thanks!

File: dask/utils.py (hunk 1 above)
Comment (jcrist) [inline dask/utils.py]: > I just want to confirm, either case works correct?

No, you need to recurse into the dispatch call here. If you don't, you return `self` (a `Dispatcher`), which is not what the user requested from calling `dispatch`. If the user is just calling the returned function then it will continue to recurse, but this isn't the intent.

File: dask/tests/test_utils.py [hunk 4]
```diff
@@ -54,11 +54,31 @@ class Bar(object):
...
     assert foo(b) == b
     assert foo((1, 2.0, b)) == (2, 1.0, b)
 
 
+def test_dispatch_lazy():
+    # this tests the recursive component of dispatch
+    foo = Dispatch()
+    foo.register(int, lambda a: a)
+
+    import numpy as np
```
Comment (jcrist) [inline dask/tests/test_utils.py]: We don't require `numpy` for this test file. You'd either need to use `pytest.importorskip('numpy')` here, or (better) use a module from the standard library. Perhaps `decimal`?

File: dask/utils.py [hunk 5]
```diff
@@ -381,31 +381,38 @@ def wrapper(func):
 
         return wrapper(func) if func is not None else wrapper
 
-    def __call__(self, arg):
-        # Fast path with direct lookup on type
+    def dispatch(self, cls):
+        """Return the function implementation for the given ``cls``"""
+        # Fast path with direct lookup on clse
```
Comment (jcrist) [inline dask/utils.py]: clse -> cls

File: dask/utils.py [hunk 6]
```diff
@@ -381,31 +381,38 @@ def wrapper(func):
...
         # Walk the MRO and cache the lookup result
-        for cls in inspect.getmro(typ)[1:]:
+        for cls in inspect.getmro(cls)[1:]:
             if cls in lk:
-                lk[typ] = lk[cls]
-                return lk[cls](arg)
-        raise TypeError("No dispatch for {0} type".format(typ))
+                lk[cls] = lk[cls]
+                return lk[cls]
+        raise TypeError("No dispatch for {0} clse".format(cls))
```
Comment (jcrist) [inline dask/utils.py]: Type `clse`. We shouldn't have abbreviations in error messages either. I'd actually just shorten this to `"no dispatch found for {0}".format(cls)`.

File: dask/tests/test_utils.py [hunk 7]
```diff
@@ -54,11 +54,34 @@ class Bar(object):
...
+    def foo_dec(a):
+        return a + 1
+
+    @foo.register_lazy("decimal")
+    def register_numpy():
+        import decimal
+        foo.register(decimal.Decimal, foo_dec)
+
+    # This test needs to be *before* any other calls
+    assert foo.dispatch(decimal.Decimal) == foo_dec
```
Comment (ordirules) [inline dask/tests/test_utils.py]: @jcrist added recursive test here

File: dask/utils.py (hunk 1 above)
Comment (ordirules) [inline dask/utils.py]: that's true, user may not necessarily call it, I added a test to that logic. thanks!

File: dask/tests/test_utils.py (hunk 4 above)
Comment (ordirules) [inline dask/tests/test_utils.py]: decimal sounds good thanks!

File: dask/utils.py (hunk 6 above)
Comment (ordirules) [inline dask/utils.py]: oops, `clse` was the result of find and replace run on `typ` -> `cls` on `type`. thanks!

File: dask/utils.py [hunk 8]
```diff
@@ -381,31 +381,38 @@ def wrapper(func):
...
             register = self._lazy.pop(toplevel)
         except KeyError:
             pass
         else:
             register()
-            return self(arg)  # recurse
+            return self.dispatch(cls) # recurse
         # Walk the MRO and cache the lookup result
-        for cls in inspect.getmro(typ)[1:]:
+        for cls in inspect.getmro(cls)[1:]:
```
Comment (jcrist) [inline dask/utils.py]: This is no longer correct, as you're shadowing `cls` when you iterate (sorry I missed catching this earlier). You need to rename the iteration variable (`cls2`?).

File: dask/tests/test_utils.py (hunk 7 above)
Comment (ordirules) [inline dask/tests/test_utils.py]: (I can also confirm it did break if I did not make the change you requested)

File: dask/tests/test_utils.py [hunk 9]
```diff
@@ -54,11 +54,34 @@ class Bar(object):
...
+        return a + 1
+
+    @foo.register_lazy("decimal")
+    def register_decimal():
+        import decimal
+        foo.register(decimal.Decimal, foo_dec)
+
+    # This test needs to be *before* any other calls
+    assert foo.dispatch(decimal.Decimal) == foo_dec
+    assert foo(decimal.Decimal(1)) == np.array(decimal.Decimal(2))
```
Comment (jcrist) [inline dask/tests/test_utils.py]: Stray call to `np.array`

Comment (ordirules) [inline dask/tests/test_utils.py]: interesting, tests passed with it. thanks!

File: dask/utils.py [hunk 10]
```diff
@@ -381,31 +381,38 @@ def wrapper(func):
...
+            return self.dispatch(cls) # recurse
         # Walk the MRO and cache the lookup result
-        for cls in inspect.getmro(typ)[1:]:
-            if cls in lk:
-                lk[typ] = lk[cls]
-                return lk[cls](arg)
-        raise TypeError("No dispatch for {0} type".format(typ))
+        for cls2 in inspect.getmro(cls)[1:]:
+            if cls2 in lk:
+                lk[cls2] = lk[cls2]
```
Comment (jcrist) [inline dask/utils.py]: This is still wrong. Should be `lk[cls] = lk[cls2]`

File: dask/utils.py (hunk 8 above)
Comment (ordirules) [inline dask/utils.py]: I should have also seen that. I am guessing since `inspect.getmro` returned its result it somehow didn't conflict with the loop. (so no test needed?)

thanks!

File: dask/utils.py (hunk 10 above)
Comment (ordirules) [inline dask/utils.py]: sorry, mindless change. I need to think about this, and why the test was even passing (if you have an idea let me know)

Comment (ordirules) [inline dask/utils.py]: (got it, I'll add test for MRO)

Comment (jcrist) [inline dask/utils.py]: This is just an optimization to avoid repeating the mro traversal for the given type. So the tests wouldn't fail, but the code wasn't doing what we wanted. I'd say just skip adding a test and make the change I requested above.

Comment (ordirules) [inline dask/utils.py]: thanks, I left it alone. forced pushed as one commit.

**Comment by jcrist** [comment]
LGTM. Will merge when tests pass.
